Add docs for vitess connector for changes made in DBZ-7698 & DBZ-7773 (#5501)
debezium/debezium: documentation/modules/ROOT/pages/connectors/vitess.adoc

}
----

[[vitess-ordered-transaction-metadata]]
=== Ordered Transaction Metadata
You can configure {prodname} to include additional metadata in data change event records.

> Such supplemental metadata can assist downstream consumers in processing messages in the correct order  when repartitioning, or some other disruption, might otherwise lead to data being consumed out of order.

Reviewer: ```suggestion
Such supplemental metadata can assist downstream consumers in processing messages in the correct order when repartitioning, or some other disruption, might otherwise lead to data being consumed out of order.
```

.Change Data Enrichment

> When transaction metadata factory is set to `VitessOrderedTransactionMetadataFactory`, the data message `Envelope` is also enriched with a new `transaction` field.

Reviewer: ```suggestion
To configure the connector to emit enriched data change event records, set the xref:vitess-property-transaction-metadata-factory[`transaction.metadata.factory`] property.
When this property is set to `VitessOrderedTransactionMetadataFactory`, the connector includes a `transaction` field in the message `Envelope`.
```

> This field provides information about every event in the form of a composite of fields. With ordered transaction metadata enabled, there are two additional fields included:

Reviewer: ```suggestion
The `transaction` field adds metadata that provides information about the order in which transactions occur.
The following fields are added to every message:
```

|===

[id="vitess-temporal-types"]
=== Temporal types

Excluding the `TIMESTAMP` data type, Vitess temporal types depend on the value of the `time.precision.mode` connector configuration property.

.Temporal values without time zones
The `DATETIME` type represents a local date and time such as "2018-01-13 09:48:27". As you can see, there is no time zone information. Such columns are converted into epoch milliseconds or microseconds based on the column’s precision by using UTC. The `TIMESTAMP` type represents a timestamp without time zone information. It is converted by MySQL from the server (or session’s) current time zone into UTC when writing and from UTC into the server (or session's) current time zone when reading back the value. For example:

* `DATETIME` with a value of `2018-06-20 06:37:03` becomes `1529476623000`.
* `TIMESTAMP` with a value of `2018-06-20 06:37:03` becomes `2018-06-20T13:37:03Z`.

> Such columns are converted into an equivalent `io.debezium.time.ZonedTimestamp` in UTC based on the server (or session’s) current time zone. The time zone will be queried from the server by default. If this fails, it must be specified explicitly by the database `connectionTimeZone` MySQL configuration option. For example, if the database’s time zone (either globally or configured for the connector by means of the `connectionTimeZone` option) is "America/Los_Angeles", the TIMESTAMP value "2018-06-20 06:37:03" is represented by a `ZonedTimestamp` with the value "2018-06-20T13:37:03Z".

Reviewer: ```suggestion
Such columns are converted into an equivalent `io.debezium.time.ZonedTimestamp` in UTC, based on the time zone of the server or session. 
By default, {prodname} queries the server for the time zone. 
If this fails, you must explicitly specify the timezone by setting the `connectionTimeZone` option in the JDBC connection string.
For example, if the database’s time zone (either globally, or as configured for the connector by means of the `connectionTimeZone` option) is "America/Los_Angeles", the TIMESTAMP value "2018-06-20 06:37:03" is represented by a `ZonedTimestamp` with the value "2018-06-20T13:37:03Z".
```

> The time zone of the JVM running Kafka Connect and Debezium does not affect these conversions.

Reviewer: ```suggestion
The time zone of the JVM that runs Kafka Connect and {prodname} does not affect these conversions.
```

> More details about properties related to temporal values are in the documentation for xref:vitess-connector-properties[Connector configuration properties].

Reviewer: ```suggestion
For more information about properties that affect temporal values, see the xref:vitess-connector-properties[connector configuration properties].
```

time.precision.mode=adaptive_time_microseconds(default)::

> The Vitess connector determines the literal type and semantic type based on the column's data type definition so that events represent exactly the values in the database. All time fields are in microseconds. Only positive `TIME` field values in the range of `00:00:00.000000` to `23:59:59.999999` can be captured correctly.

Reviewer: ```suggestion
The Vitess connector determines the literal type and semantic type based on the column's data type definition so that events represent exactly the values in the database. 
All time fields are in microseconds. 
Only positive `TIME` field values in the range of `00:00:00.000000` to `23:59:59.999999` can be captured correctly.
```

+
.Mappings when `time.precision.mode=adaptive_time_microseconds`
[cols="25%a,20%a,55%a",options="header",subs="+attributes"]
|===
|Vitess type |Literal type |Semantic type

|`DATE`
|`INT32`
a|`io.debezium.time.Date` +

> Represents the number of days since the epoch.

Reviewer: ```suggestion
Represents the number of days elapsed since the UNIX epoch.
```

|`TIME[(M)]`
|`INT64`
a|`io.debezium.time.MicroTime` +

> Represents the time value in microseconds and does not include time zone information. MySQL allows `M` to be in the range of `0-6`.

Reviewer: ```suggestion
Represents the time value in microseconds and does not include time zone information. 
MySQL allows `M` to be in the range of `0-6`.
```

|`DATETIME, DATETIME(0), DATETIME(1), DATETIME(2), DATETIME(3)`
|`INT64`
a|`io.debezium.time.Timestamp` +

> Represents the number of milliseconds past the epoch and does not include time zone information.

Reviewer: ```suggestion
Represents the number of milliseconds elapsed since the UNIX epoch and does not include time zone information.
```

|`DATETIME(4), DATETIME(5), DATETIME(6)`
|`INT64`
a|`io.debezium.time.MicroTimestamp` +

> Represents the number of microseconds past the epoch and does not include time zone information.

Reviewer: ```suggestion
Represents the number of microseconds elapsed since the UNIX epoch and does not include time zone information.
```

|===

time.precision.mode=connect::

> The Vitess connector uses defined Kafka Connect logical types. This approach is less precise than the default approach and the events could be less precise if the database column has a _fractional second precision_ value of greater than `3`. Values in only the range of `00:00:00.000` to `23:59:59.999` can be handled. Set `time.precision.mode=connect` only if you can ensure that the `TIME` values in your tables never exceed the supported ranges. The `connect` setting is expected to be removed in a future version of {prodname}.

Reviewer: ```suggestion
The Vitess connector uses defined Kafka Connect logical types. 
This approach is less precise than the default approach, and the events could be less precise if the database column has a _fractional second precision_ value that is greater than `3`. 
The connector can process values that range from `00:00:00.000` to `23:59:59.999`. 
Set `time.precision.mode=connect` only if you are certain that the `TIME` values in your tables never exceed the supported ranges. 
The `connect` setting is expected to be removed in a future version of {prodname}.
```

+
.Mappings when `time.precision.mode=connect`
[cols="25%a,20%a,55%a",options="header",subs="+attributes"]
|===
|Vitess data type |Literal type |Semantic type

|`DATE`
|`INT32`
a|`org.apache.kafka.connect.data.Date` +

> Represents the number of days since the epoch.

Reviewer: ```suggestion
Represents the number of days elapsed since the UNIX epoch.
```

|`TIME[(M)]`
|`INT64`
a|`org.apache.kafka.connect.data.Time` +
Represents the time value in microseconds since midnight and does not include time zone information.

|`DATETIME[(M)]`
|`INT64`
a|`org.apache.kafka.connect.data.Timestamp` +

> Represents the number of milliseconds since the epoch, and does not include time zone information.

Reviewer: ```suggestion
Represents the number of milliseconds elapsed since the UNIX epoch, and does not include time zone information.
```

|===

[id="vitess-temporal-types"]
=== Temporal types

Excluding the `TIMESTAMP` data type, Vitess temporal types depend on the value of the xref:vitess-property-time-precision-mode[`time.precision.mode`] connector configuration property.

.Temporal values without time zones
The `DATETIME` type represents a local date and time such as "2018-01-13 09:48:27".
As you can see in the preceding example, this type does not include time zone information.
Columns of this type are converted into epoch milliseconds or microseconds based on the column’s precision by using UTC.
The `TIMESTAMP` type represents a timestamp without time zone information.
When writing data, MySQL converts the `TIMESTAMP` type from the time zone of the server or session into UTC format.

> When it reads values, the database converts from UTC format to the current time zone of the  server or session.

Reviewer: ```suggestion
When it reads values, the database converts from UTC format to the current time zone of the server or session.
```

`never`::
When the connector starts, it skips the snapshot process and immediately begins to stream change events for operations that the database records to the binary logs.

|[[vitess-property-time-precision-mode]]<<vitess-property-time-precision-mode, `+time.precision.mode+`>>
|`adaptive_time_microseconds`

> |Time, date, and timestamps can be represented with different kinds of precision, including: +

Reviewer: ```suggestion
|You can set the following options to determine how {prodname} represents the precision of time, date, and timestamps values: +
```

+

> `adaptive_time_microseconds` (the default) captures the date, datetime and timestamp values exactly as in the database using either millisecond, microsecond, or nanosecond precision values based on the database column's type, with the exception of TIME type fields, which are always captured as microseconds. +
> +
> `connect` always represents time and timestamp values using Kafka Connect's built-in representations for Time, Date, and Timestamp, which use millisecond precision regardless of the database columns' precision.

Reviewer: ```suggestion
`adaptive_time_microseconds`:: 
(Default) Captures date, datetime, and timestamp values exactly as they exist in the database. 
Values are represented with a precision in milliseconds, microseconds, or nanoseconds, depending on the database column type, with the exception of `TIME` type fields, which are always captured as microseconds. 

`connect`:: 
Time and timestamp values are always represented in the default Kafka Connect formats for Time, Date, and Timestamp, which use millisecond precision regardless of the database columns' precision.
```

`never`::
When the connector starts, it skips the snapshot process and immediately begins to stream change events for operations that the database records to the binary logs.

|[[vitess-property-time-precision-mode]]<<vitess-property-time-precision-mode, `+time.precision.mode+`>>
|`adaptive_time_microseconds`

> |You can set the following options to determine how {prodname} represents the precision of time, date, and timestamps values: +

Reviewer: ```suggestion
|You can set the following options to determine how {prodname} represents the precision of time, date, and timestamp values: +
```

|`false`
|Determines whether the connector generates events with transaction boundaries and enriches change event envelopes with transaction metadata. Specify `true` if you want the connector to do this. See xref:vitess-transaction-metadata[Transaction metadata] for details.

|[[vitess-property-transaction-metadata-factory]]<<vitess-property-transaction-metadata-factory, `transaction.metadata.factory`>>
|`io.debezium.pipeline.txmetadata.DefaultTransactionMetadataFactory`

> |Determines the class used to track transaction context and build transaction struct & schemas.

Reviewer: ```suggestion
|Determines the class that the connector uses to track transaction context and build the data structures and schemas to represent transactions.
```

> `io.debezium.connector.vitess.pipeline.txmetadata.VitessOrderedTransactionMetadataFactory` will provide additional transaction metadata for establishing ordering between two events independent of the order they are consumed. See xref:vitess-ordered-transaction-metadata[Ordered transaction metadata] for details.

Reviewer: ```suggestion
`io.debezium.connector.vitess.pipeline.txmetadata.VitessOrderedTransactionMetadataFactory` provides additional transaction metadata that can help consumers to interpret the correct order of two events, regardless of the order in which they are consumed. 
For more information, see xref:vitess-ordered-transaction-metadata[Ordered transaction metadata].
```
